src/main/java/com/fasterxml/jackson/core/util/BufferRecyclerPool.java

```diff
@@ -1,17 +1,157 @@
+import java.util.concurrent.atomic.AtomicReference;
 
 /**
  * Interface for entity that controls creation and possible reuse of {@link BufferRecycler}
  * instances used for recycling of underlying input/output buffers.
  *
  * @since 2.16
  */
-public interface BufferRecyclerPool
-    extends java.io.Serializable
-{
-    public BufferRecycler acquireBufferRecycler(TokenStreamFactory forFactory);
+public interface BufferRecyclerPool extends Serializable {
 
-    public void releaseBufferRecycler(BufferRecycler recycler);
+    BufferRecycler acquireBufferRecycler();
+
+    void releaseBufferRecycler(BufferRecycler recycler);
+
+    /**
+     * Returns the default BufferRecyclerPool implementation which is the thread local based one.
+     */
+    static BufferRecyclerPool defaultRecyclerPool() {
+        return PoolStrategy.THREAD_LOCAL.getPool();
+    }
+
+    enum PoolStrategy {
+        NO_OP(new NonRecyclingRecyclerPool()),
+        THREAD_LOCAL(new ThreadLocalRecyclerPool()), 
+        CONCURRENT_DEQUEUE(new ConcurrentDequePool()),
+        LOCK_FREE(new LockFreePool());
+
+        private final BufferRecyclerPool pool;
+
+        PoolStrategy(BufferRecyclerPool pool) {
+            this.pool = pool;
+        }
+
+        public BufferRecyclerPool getPool() {
+            return pool;
+        }
+    }
+
+    /**
+     * Default {@link BufferRecyclerPool} implementation that uses
+     * {@link ThreadLocal} for recycling instances.
+     *
+     * @since 2.16
+     */
+    class ThreadLocalRecyclerPool implements BufferRecyclerPool {
+        private static final long serialVersionUID = 1L;
+
+        @Override
+        public BufferRecycler acquireBufferRecycler() {
+            return BufferRecyclers.getBufferRecycler();
+        }
+
+        @Override
+        public void releaseBufferRecycler(BufferRecycler recycler) {
+            ; // nothing to do, relies on ThreadLocal
+        }
+    }
+
+    /**
+     * {@link BufferRecyclerPool} implementation that does not use
+     * any pool but simply creates new instances when necessary.
+     *
+     * @since 2.16
+     */
+    class NonRecyclingRecyclerPool implements BufferRecyclerPool {
+        private static final long serialVersionUID = 1L;
+
+        @Override
+        public BufferRecycler acquireBufferRecycler() {
+            return new BufferRecycler();
+        }
+
+        @Override
+        public void releaseBufferRecycler(BufferRecycler recycler) {
+            ; // nothing to do, relies on ThreadLocal
+        }
+    }
+
+    /**
+     * {@link BufferRecyclerPool} implementation that uses
+     * {@link ConcurrentLinkedDeque} for recycling instances.
+     *
+     * @since 2.16
+     */
+    class ConcurrentDequePool implements BufferRecyclerPool {
+        private final Deque<BufferRecycler> pool = new ConcurrentLinkedDeque<>();
+
+        @Override
+        public BufferRecycler acquireBufferRecycler() {
+            return getBufferRecycler().withPool(this);
```

Comment:
Instead of chaining like this, could add a new constructor in `BufferRecycler`. Although if ref is to be moved that won't be necessary.

Reply:
I will need this method anyway because the pool from where this `BufferRecycler` has been taken will have to be set not only when the instance is created for the first time, but also when an existing one is taken again from the pool and acquired by a different  `IOContext`. That because when the `BufferRecycler` is released to the pool I nullify the reference to the pool itself as also suggested by @franz1981.

Reply:
Ah. Yes, right you are.